conformance/clitest.go

```diff
@@ -0,0 +1,143 @@
+// Package conformance is a test suite for testing CLI parsing into jsonnext
+// data structures. It can be used to check conformance of a CLI parsing
+// implementation.
+package conformance
+
+import (
+	"strings"
+	"testing"
+
+	"foxygo.at/s/test"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	"foxygo.at/jsonnext"
+)
+
+// CLIParser should be implemented by a jsonnext.Config CLI parsing
+// implementation to test its conformance to the standard parsing expected.
+// Parse() is called with a given command line for which the parse must return
+// a *jsonnext.Config corresponding to that set of CLI flags or return an error
+// where those flags fail to parse..
+type CLIParser interface {
+	Parse(t *testing.T, args []string) (*jsonnext.Config, error)
+}
+
+// Suite is a conformance test for command line parsers that parse into a
+// jsonnext.Config struct. The test case for the parser under test must
+// implement CLIParser to parse a given command line into a *jsonnext.Config
+// struct.
+type Suite struct {
+	suite.Suite
+	parser CLIParser
+}
+
+// NewSuite returns a new Suite to test the conformance of the given parser.
+func NewSuite(s CLIParser) *Suite {
+	return &Suite{parser: s}
+}
+
+// Run runs the conformance test in this package against the CLIParser
+// configured in the Suite.
+func (s *Suite) Run(t *testing.T) {
+	suite.Run(t, s)
+}
+
+var flags = map[string]struct {
+	flag    string
+	makevar func(string) jsonnext.VMVar
+}{
+	"ext-str":       {"--ext-str", jsonnext.NewExtStr},
+	"ext-str-short": {"-V", jsonnext.NewExtStr},
+	"ext-code":      {"--ext-code", jsonnext.NewExtCode},
+	"ext-str-file":  {"--ext-str-file", jsonnext.NewExtStrFile},
+	"ext-code-file": {"--ext-code-file", jsonnext.NewExtCodeFile},
+	"tla-str":       {"--tla-str", jsonnext.NewTLAStr},
+	"tla-str-short": {"-A", jsonnext.NewTLAStr},
+	"tla-code":      {"--tla-code", jsonnext.NewTLACode},
+	"tla-str-file":  {"--tla-str-file", jsonnext.NewTLAStrFile},
+	"tla-code-file": {"--tla-code-file", jsonnext.NewTLACodeFile},
+}
+
+// TestVMVars tests that all styles of VMVars are correctly parsed on the
+// command line, either as literals on the command line or from the
+// environment.
+func (s *Suite) TestVMVars() {
+	for name, tc := range flags {
+		name, tc := name, tc
+		s.T().Run(name, func(t *testing.T) {
```

Comment:
To me this is fine, just wanted to remind you that in a different place you said this should be:
`t := s.T()` as the first line of the function and then use `t` as we are used to...

Reply:
I can't remember what I was thinking when I wrote that, but it looks to me that I did it this way because it is used once and not as a parameter. There is also a local `t` in the closure, so we don't use a different `t` in two places in the function.